### ISSUE TITLE
Support `Box` overflow and scrolling

### ISSUE THREAD

**Issue by gnidan** [vadimdemedes/ink#432]
Firstly: this library is amazing; I'm so excited to use it everywhere I possibly can. Sadly, this issue describes a use case that's blocking me from moving forward :(

I'm opening this issue and offering a corresponding [Gitcoin](https://gitcoin.co/) bounty in hopes it will help. Thank you to anyone who can help make this happen!

## Summary

This feature request seeks to cover the use case where a component may exceed the bounds of its container. Specifically, this issue seeks the following behaviors:
- Allow `<Box overflow={...}>`, perhaps as implemented by #393 
- Allow Box components to scroll their contents. Based on my research, #393 alone is insufficient, since it doesn't provide a way for a component to measure what its full width/height would be. (When `overflow="hidden"` is provided, width/height are measured based on the final component on-screen, and there seems to be no way around this)

## Related

PR #393 (Add `overflow` prop to Box component) is mostly complete for my concerns here. I outlined my concerns with scrolling / measurements in [this comment](https://github.com/vadimdemedes/ink/pull/393#issuecomment-798139937), but no idea what the right answer is.

https://github.com/vadimdemedes/goodness-squad/issues/5 describes a related use case (Fullscreen UIs). This issue doesn't target fullscreen UIs specifically; overflow/scrolling behavior should work regardless of whether the component is going to be fullscreen or not.

## Acceptance criteria

See [comment below](https://github.com/vadimdemedes/ink/issues/432#issuecomment-821921475)

## Notes

Please let me know if the requirements could use clarification! I'm very keen to get this working 🙇

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1.0 ETH (2414.48 USD @ $2414.48/ETH)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.com/invite/83BK5z55yB/'>Gitcoin's Discord</a>
* $2,187,764.65 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by chrisirhc** [comment]
> Allow Box components to scroll their contents.
Do you mean the box will scroll the contents over time? Such as an animation?

Based on what you're saying sounds like you might need to:
1. Render the inner box contents then measure it's height and width
2. Now set the height and width explicitly and render it within the Box and scroll it.

This is assuming you don't know the inner box contents from the beginning, since it could be rendered via flexbox rules.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 265 years, 7 months from now.
Please review their action plans below:


**1) [sunguru98](https://gitcoin.co/sunguru98) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?mutate_worker_action=approve&worker=sunguru98) | [reject worker](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?mutate_worker_action=reject&worker=sunguru98))_.**

Hey there, I am a React developer for almost 4 years, and this seems to be a challenging problem for me to figure out. There is a way which I am thinking would be possible, but I really would like to take this task to see what could be possible. 

You can check out my GitHub @ https://github.com/sunguru98 
Looking forward to figure this out. Thanks :)

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559).

**Comment by chrisirhc** [comment]
Here's a solution based on #393 : https://gist.github.com/chrisirhc/bf3f973e5d4db2cbc96d95d1f023f0ad

**Comment by gnidan** [comment]
@chrisirhc I don't see where your solution measures the full size of the inner contents... it seems like this only scrolls a known amount. I really want to be able to implement an up/down pager (think `less`), but the problem I run into is that there's no way of knowing when we've scrolled to the bottom or top. Here's what I'm doing:

```typescript
import React from "react";
import { Box, Text } from "ink";

import { loremIpsum } from "lorem-ipsum";

export const Paragraphs = ({ count = 10 }) => {
  const raw = loremIpsum({
    units: "paragraphs",
    count
  });

  const text = raw.split("\n").join("\n\n");

  return (
    <Box>
      <Text>{text}</Text>
    </Box>
  )
}

export const Main = () => {
  return (
    <Box width={80} height={25}>
      <AmazeBoxRenderer>
        <Box width={80}>
          <Paragraphs count={10} />
        </Box>
      </AmazeBoxRenderer>
    </Box>
  );
}
```

When I run this, I see `80 25` logged, which seems to suggest that the renderer believes the full height of the content to be merely 25 lines. This isn't the case, since ten paragraphs of lorem ipsum exceeds that at 80 characters. Does this break your assumptions here?

To get to your other points,

> Based on what you're saying sounds like you might need to:
> 
> 1. Render the inner box contents then measure it's height and width
> 2. Now set the height and width explicitly and render it within the Box and scroll it.

I'm not sure this works, since Ink's renderer will respect the outer wrapper's bounds rather than letting some inner content render with whatever natural dimensions it has.

> This is assuming you don't know the inner box contents from the beginning, since it could be rendered via flexbox rules.

Definitely want to maintain this assumption.

I'll write another comment to follow-up with some acceptance criteria, since I've thought of it. Regardless, thanks for taking an initial look @chrisirhc! Tricky problem, right?

**Comment by gnidan** [comment]
## Acceptance criteria

- **A1.** Given component _A_ of unknown "natural size" _(x<sub>A</sub>, y<sub>A</sub>)_ inside parent component _B_ of fixed size _(x<sub>B</sub>, y<sub>B</sub>)_,

  - **A1.1.** _A_ should render within the bounds of _B_ with overflow hidden
 
  - **A1.2.** Either _A_ or _B_ should be able to compute _A_'s natural size _(x<sub>A</sub>, y<sub>A</sub>)_, e.g. to ensure that _A_ never completely leaves the viewport of _B_.

- **A2.** Solutions must perform reasonably well without noticeable re-rendering or other flickering. This might mean fixing the flickering described in https://github.com/vadimdemedes/ink/issues/359, but not necessarily (since A1 paired with stdout measurement means it'd be trivial to prevent rendering something larger than the terminal).

- **A3.** Solutions must work with existing components in the ecosystem, including [ink-spinner](https://github.com/vadimdemedes/ink-spinner), [ink-select-input](https://github.com/vadimdemedes/ink-select-input), [ink-progress-bar](https://github.com/brigand/ink-progress-bar), etc. This means one of two options:

  - **A3 (option 1, _strongly preferred_).** Solution is merged and released as part of this library's official distribution channel. This means getting @vadimdemedes's approval, but I don't want to rely on that for this bounty.
 
  - **A3 (option 2).** Solution is releasable as a forked NPM package. Two stipulations:
 
    - **A3(2).1.** Changes are opened as PR here (so that I may seek to champion getting the changes into a main-line release once your work is done)
  
    - **A3(2).2.** Solution includes clear instructions on getting the above-mentioned component packages to work cleanly. Those packages use `"ink"` itself as a dependency in `package.json`, and that means ink-spinner (e.g.) will try to use the officially-released rendering logic as opposed to whatever modified version your solution provides. (I ran into this problem when I released #393 as a fork.) Solutions that require forking any/all component packages are not acceptable, and my use case doesn't afford leveraging systems like yarn's [`resolutions` config](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/). (I intend to use this work in a dependent application released to NPM and installable with `npm install`, so yarn-specific niceties are sadly just not going to cut it.)

- **A4.** Solution includes automated tests and follows consistent style and quality alongside the rest of Ink's codebase.

This is about the extent of what I can think of as necessary, but of course I understand that there may be ambiguity or technical limitations when it comes to the details. Please don't hesitate to communicate any comments, questions, or concerns.

Thank you to everyone who has thus far applied to work on this bounty! I will continue reviewing each of your applications and will respond to each individually in the next 48 hours. ❤️

**Comment by gitcoinbot** [comment]
@developerfred Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?snooze=1) | [3 days](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?snooze=3) | [5 days](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?snooze=5) | [10 days](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?snooze=10) | [100 days](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559?snooze=100)

**Comment by developerfred** [comment]
Hi @gitcoinbot my code will be here https://replit.com/@codingsh/poc-ink-box-upgrade

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work for 1.0 ETH (2311.76 USD @ $2425.56/ETH) has been submitted by__:

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/vadimdemedes/ink/432/100025559).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.com/invite/83BK5z55yB/'>Gitcoin's Discord</a>
* $2,166,350.12 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by developerfred** [comment]
@gnidan there is a way for me to get the height of the elements with this hook, I will write more test and improve the poc.

**Comment by gnidan** [comment]
Awesome @developerfred, glad to hear about progress!

In the meantime, I grabbed the relevant stuff from my use case and made a repo for it at [gnidan/ink-scroller](https://github.com/gnidan/ink-scroller). There's something strange going on with left/right scrolling, but I'm less focused on that, since I just haven't looked into it.

Point of that repo: I want to be able to stop scrolling at the end of the text :)

**Comment by dustinlacewell** [comment]
![2021-07-08-22-50-43](https://user-images.githubusercontent.com/53952/125000135-56323c80-e015-11eb-9ee6-6247651f14da.gif)

**Comment by navxio** [comment]
@developerfred any luck? Can I offer (free) help?

**Comment by developerfred** [comment]
@navxio sure!

**Comment by vadimdemedes** [comment]
I think it would be interesting to have that as a separate npm package, but I'm not sure I'd want to have that in Ink itself. If you have a complicated CLI that requires scrolling areas, you might be better of with ncurses-based alternative like [blessed](https://github.com/chjj/blessed).

**Comment by zach-is-my-name** [comment]
> I think it would be interesting to have that as a separate npm package, but I'm not sure I'd want to have that in Ink itself. If you have a complicated CLI that requires scrolling areas, you might be better of with ncurses-based alternative like [blessed](https://github.com/chjj/blessed).

The issue with blessed and similar is we don't get the declarative dom-like environment that we get here... as it stands and (please someone correct me if I'm wrong), Ink is the only project that makes building a CLI or TUI more or less like building for the web... 

And I believe that's hugely important because we've come to realize the terminal offers one of the most productive, distraction free environments in existence... a turn-key solution to scrolling is absolutely fundamental in my opinion. Thanks for listening

**Comment by gnidan** [comment]
Glad to see some activity on this issue. I'll review the acceptance criteria and state of current efforts. I still want this, and I'm still willing to spend the listed bounty!

Thanks for your insight @vadimdemedes and @zach-is-my-name 🙏

**Comment by gnidan** [comment]
> I think it would be interesting to have that as a separate npm package, but I'm not sure I'd want to have that in Ink itself.

@vadimdemedes I remember from my exploration into this issue that we can't get a separate NPM package to work without a change to Ink's rendering methods... since Ink is responsible for doing all the layout math, it needs to keep track of the "full size" dimensions as well as the computed dimensions on-screen (what I call "natural size" vs. "fixed size" above), but the problem is that there's no way (right now) to create a react hook to report the natural size dimensions. It looks like this information gets lost somewhere deep inside the renderer (or worse, inside the flowbox computations).

This stuff is all very confusing to me, however, since flowbox doesn't (didn't?) have good API documentation. Maybe I'm wrong.


> Hi @gitcoinbot my code will be here https://replit.com/@codingsh/poc-ink-box-upgrade

@developerfred is this still current? It seems rather barebones, and it doesn't currently even run at all (looks like you have a syntax error).

---

Anyway, I notice there is some movement on #412 and also I see #472 is new since I've spent effort investigating the state of this issue. @vadimdemedes could you please comment on what a path forward might look like? I'm fine if it is actually possible to achieve this "natural size" calculation in a separate NPM package, but I just don't see how that's possible.

If further coordination across existing PRs is necessary, I could potentially double the bounty and split it up across contributors/maintainers, since I _really want_ this behavior, but mind you I'm only working with my personal funds here. But if this is a multi-person effort, I do want to recognize that and reward everyone's efforts. This behavior is the only thing stopping me from using Ink for a number of projects of varying sizes, and I can't move forward on those projects in a way that satisfies me (because, well, who wants to use blessed when ink exists ;)

**Comment by zach-is-my-name** [comment]
#521 what say you @gnidan?

**Comment by gnidan** [comment]
This is great! Thanks for getting back to this @vadimdemedes. I just adapted your example for my old https://github.com/gnidan/ink-scroller repo and I am able to get up/down scrolling to work perfectly, with limits for when the end of content gets reached. (The `Math.max()` business seems a bit prone to off-by-one errors, but I think that's just the way I have it set up. Not a real issue, I don't think.)

I'm not super concerned with the focus-based scrolling, since I think there's a relatively straightforward solution to check for own focus, plus pass some `handleChildFocus` function as a prop to the children (or similar). It's a bit hacky, but anything more robust would probably require Ink to do more heavy-lifting, and that seems excessive for now. Maybe in the future you'll want to address this limitation? This limitation is fine for me right now, at least.

One question, since my example ink-scroller repo uses random paragraphs of lorem-ipsum text. When I scroll left/right, Ink re-renders the paragraphs every time I scroll (i.e., it computes the word wrapping every scroll action, which makes individual lines re-wrap differently).

<details>
<summary>See GIF of what I mean</summary>
 
![scroller](https://user-images.githubusercontent.com/151065/234746194-02568431-dc77-4ddc-9620-9503ca0cddc0.gif)

</details>

It's possible that I just have a bug, since it's been awhile since I've looked at my example code. But in case you might know... should I just plan to add line-breaks manually and not rely on Ink's wrapping logic?

**Comment by gnidan** [comment]
Regarding the bounty portion of this issue, I consider the matter resolved. I can use the regular `"ink"` NPM package to get `<Box />` with both overflow support and support for bounded scrolling, which is exactly what I have needed!

With pleasure, I'd like to remit the 1 ETH funds associated with this bounty. 

@developerfred: Although your solution did not meet acceptance criteria, I'm [sending](https://etherscan.io/tx/0x19d1959f31444aed8b2b8620b7d8ed70849647dc94a459d7d68491087549ca2d) you a 0.1 ETH tip for your attempt when I first opened this. You're already in the Gitcoin system, so I was able to use the Gitcoin webapp UI to issue this payout... unfortunately I didn't check the gas price, so that transaction may get dropped. Don't worry; I will monitor this and make sure you get paid.

@tin-t: I'd like to offer you a 0.2 ETH tip as gratitude for your work on the initial implementation of `<Box overflow />` support in #393.

@vadimdemedes: Between #502 and https://github.com/vadimdemedes/ink/issues/432#issuecomment-1519671092, your work clearly meets the acceptance criteria, and thus you deserve the 1.0 ETH! I would like to send this to you.

But, for both @tin-t and @vadimdemedes... I don't know how to issue you these payments.

- It might be easiest if you make https://bounties.gitcoin.co/ accounts and sign up to work on this bounty (https://bounties.gitcoin.co/issue/25559). It looks like it's pretty easy for me to send arbitrary amounts to anyone who is in their system and associated with this bounty.
- Alternatively, just send me your Ethereum address (you can paste here in a comment, send it to [me](https://twitter.com/gnidan) in a Twitter DM, or we can find another means of communicating.) I can just remit the funds directly and figure out Gitcoin's commission separately.

I understand that the two of you might not be prepared to receive ETH. I apologize if this is the case, but unfortunately (for budgeting and tax reasons), I can't issue the payout in a different currency other than ETH. I am happy to walk you through the process of making an Ethereum account and helping you find a reputable fiat off-ramp that you can use if you want to convert the funds into your local currencies. Please let me know.

For housekeeping purposes, I'll plan to leave both this GH issue and the associated bounty open until it's fully paid. I worry that something on the Gitcoin side will break otherwise...

Anyway, thank you everyone who contributed to this issue, and enormous thanks to you @vadimdemedes for all your work on Ink, really :). I'm looking forward to picking up where I left off when I first got blocked by this overflow/scrolling thing.

Cheers all!

**Comment by vadimdemedes** [comment]
@gnidan That's very generous of you, thank you! If you could send it either to `0xa1b1bbB8070Df2450810b8eB2425D543cfCeF79b` or `0x93Bda139023d582C19D70F55561f232D3CA6a54c`, that'd be awesome 💛

> One question, since my example ink-scroller repo uses random paragraphs of lorem-ipsum text. When I scroll left/right, Ink re-renders the paragraphs every time I scroll (i.e., it computes the word wrapping every scroll action, which makes individual lines re-wrap differently).

I think it can be solved by setting a fixed width to the `Box` that wraps the children (not the root `Box`).

**Comment by gnidan** [comment]
> @gnidan That's very generous of you, thank you! If you could send it either to `0xa1b1bbB8070Df2450810b8eB2425D543cfCeF79b` or `0x93Bda139023d582C19D70F55561f232D3CA6a54c`, that'd be awesome 💛

[Sent](https://etherscan.io/tx/0x5aa011e475d62fe192dcd89007733e865437e98d573a6fdd4b2ffcdf86e05218). I hope it makes a difference!

> I think it can be solved by setting a fixed width to the `Box` that wraps the children (not the root `Box`).

This works, thank you!

> For housekeeping purposes, I'll plan to leave both this GH issue and the associated bounty open until it's fully paid.

I closed the bounty on the Gitcoin side, and I'm going to close this issue now.

(@tin-t it looks like you have not been active on Github in awhile, but if you see these notifications, please find a way to reach out to me so I can make good on my stated offer above! Hopefully I will see any notifications here, but if not, my username is `@gnidan` on most social media platforms)

**Comment by vadimdemedes** [comment]
Thanks @gnidan, every dollar makes the Ukrainian victory closer 💛

**Comment by janis-me** [comment]
The solution works well and I'd like to thank you all for the nice work. Made it easy for me.

Just for future readers: If you're looking for a typescript version that can re-scale based on resize events, see https://gist.github.com/janis-me/1a55a8747f12e1c4fc86ebe2d05a0a55

Combine that with `useStdoutDimensions` and it works very well!

**Comment by marbemac** [comment]
@janis-me any chance you could update the gist link? goes to a 404 atm. thanks!

**Comment by janis-me** [comment]
> @janis-me any chance you could update the gist link? goes to a 404 atm. thanks!

Ah, sure! I changed my username in the meantime. Updated it now! ([Or click here](https://gist.github.com/janis-me/1a55a8747f12e1c4fc86ebe2d05a0a55))
